openedx/core/djangoapps/credit/signals.py

```diff
@@ -70,3 +99,264 @@ def listen_for_grade_calculation(sender, username, grade_summary, course_key, de
+    """ Create verification user partition for given block.
+
+    Group and UserPartition id's will be int.
+
+    Args:`
+        course_key (CourseKey): Identifier for the course
+        user_partition (UserPartition): UserPartition object
+        block (XBlock): XBlock mixin
+
+    Returns:
+        UserPartition object.
+
+    """
+    course = modulestore().get_course(course_key, depth=0)
+
+    # make int id for user partition scheme `VerificationPartitionScheme` which
+    # is unique in user partitions of the provided course
+    group_configuration_id = generate_int_id(MINIMUM_GROUP_ID, MYSQL_MAX_INT, get_course_partitions_used_ids(course))
+    group_configuration_name = u"Verification Checkpoint for {checkpoint_name}".format(
+        checkpoint_name=block.display_name
+    )
+    group_configuration_description = group_configuration_name
+    # Since Group requires id to be an int so map verification partition group names to int
+    # 0: VerificationPartitionScheme.NON_VERIFIED
+    # 1: VerificationPartitionScheme.VERIFIED_ALLOW
+    # 2: VerificationPartitionScheme.VERIFIED_DENY
+    group_configuration_groups = [
+        Group(VerificationPartitionScheme.NON_VERIFIED, 'Not enrolled in a verified track'),
+        Group(VerificationPartitionScheme.VERIFIED_ALLOW, 'Enrolled in a verified track and has access'),
+        Group(VerificationPartitionScheme.VERIFIED_DENY, 'Enrolled in a verified track and does not have access'),
+    ]
+    group_configuration_parameters = {'location': unicode(block.location)}
+
+    group_configuration = UserPartition(
+        id=group_configuration_id,
+        name=group_configuration_name,
+        description=group_configuration_description,
+        groups=group_configuration_groups,
+        scheme=user_partition,
+        parameters=group_configuration_parameters,
+    )
+    return group_configuration
+
+
+def _get_credit_xblocks_for_access_control(course_key):  # pylint: disable=invalid-name
+    """ Retrieve all credit requirements XBlocks in the course for categories
+     in list 'GATED_CREDIT_XBLOCK_CATEGORIES'.
+
+    Args:
+        course_key (CourseKey): Identifier for the course
+
+    Returns:
+        List of XBlocks that are published and haven't been deleted.
+
+    """
+    credit_blocks = []
+    for category in GATED_CREDIT_XBLOCK_CATEGORIES:
+        xblocks = get_course_xblocks(course_key, category)
+        credit_blocks.extend(xblocks)
+
+    return credit_blocks
+
+
+def _update_published_block(block):
+    """ Update the provided XBlock in modulestore for published branch only.
+
+    Use the xblock's last editor as modifier.
+
+    Args:
+        block (XBlock): XBlock mixin
+
+    Returns:
+        List of XBlocks that are published and haven't been deleted.
+
+    """
+    # Update the published branch only
+    with modulestore().branch_setting(ModuleStoreEnum.Branch.published_only):
```

Comment:
From our conversations with @doctoryes it sounds like one potential solution would be to:

1) Call `update_item` on the draft branch of the block to update its `group_access` field.
2) Call `update_item` on the publish branch of the block to update its `group_access` field.

From what I can see by looking through the split Mongo course structure, the `has_changes` check  will compare changed fields (including `group_access`) and return `True` if it detects changes.  This is the behavior we want.

We will need to test this, but it seems like the right direction.  We definitely need this to work in Split Mongo, but we don't necessarily need to make it work in old Mongo since all the courses that will use this have been created since we switched to Split.